### PR TITLE
Issue 1340

### PR DESCRIPTION
**Pull request by jarosik10**
for: https://github.com/openstax/unified/issues/1340
There is some redundant code in `ShowPracticeQuestions ` component that probably should be fixed.

### CODE REVIEW

File: src/app/content/practiceQuestions/components/FinalScreen.tsx [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+      <NextSectionMessage
+        nextSection={nextSection}
+        messageKey='i18n:practice-questions:popup:empty:next-section'
+        onClick={() => dispatch(setSelectedSection(nextSection))}
+      />
+    }
+  </StyledFinalScreen>
+}
+
+export default FinalScreen;
```
Comment (PiotrKozlowski) [inline src/app/content/practiceQuestions/components/FinalScreen.tsx:51]: Files should end with a newline

File: src/app/content/practiceQuestions/components/FinalScreen.tsx [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+  ${textRegularStyle}
+`;
+
+// tslint:disable-next-line: variable-name
+const StyledText = styled.span`
+  margin-bottom: 3rem;
+`;
+
+interface EmptyScreenProps {
+  nextSection: LinkedArchiveTreeSection | undefined;
```
Comment (PiotrKozlowski) [inline src/app/content/practiceQuestions/components/FinalScreen.tsx:28]:
```suggestion
  nextSection?: LinkedArchiveTreeSection
```

File: src/app/content/practiceQuestions/components/FinalScreen.tsx [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+  text-align: center;
+  ${textRegularStyle}
+`;
+
+// tslint:disable-next-line: variable-name
+const StyledText = styled.span`
+  margin-bottom: 3rem;
+`;
+
+interface EmptyScreenProps {
```
Comment (PiotrKozlowski) [inline src/app/content/practiceQuestions/components/FinalScreen.tsx:27]:
```suggestion
interface FinalScreenProps {
```

File: src/app/content/practiceQuestions/components/Question.tsx [hunk 4]
```diff
@@ -64,9 +66,12 @@ const Question = () => {
   const showCorrect = showCorrectState === question;
 
   const onSubmit = (e: React.FormEvent) => {
-    // TODO: Add support for handling Finish button
     e.preventDefault();
     dispatch(setAnswer({ answer: selectedAnswer, questionId: question.uid }));
+    // handling Finish button
+    if (isFinalQuestion && isSubmitted) {
+      dispatch(finishQuestions());
+    }
```
Comment (PiotrKozlowski) [inline src/app/content/practiceQuestions/components/Question.tsx:74]: In this case, `setAnswer` will be called after clicking `Finish`
This statement should be probably directly after the `e.preventDefault();` and `return`

File: src/app/content/practiceQuestions/components/QuestionNavigation/index.tsx [hunk 5]
```diff
@@ -42,7 +42,11 @@ const QuestionNavigation = ({ question, selectedAnswer, ...props }: QuestionNavi
             onClick={(e: React.MouseEvent) => {
               e.preventDefault();
               dispatch(setAnswer({ answer: null, questionId: question.uid }));
-              dispatch(nextQuestion());
+              if (!isFinalQuestion) {
+                dispatch(nextQuestion());
+              } else {
+                dispatch(finishQuestions());
+              }
```
Comment (PiotrKozlowski) [inline src/app/content/practiceQuestions/components/QuestionNavigation/index.tsx:49]: It will be simpler without negation
```suggestion
              if (isFinalQuestion) {
                dispatch(finishQuestions());
              } else {
                dispatch(nextQuestion());
              }
```

File: src/app/content/practiceQuestions/components/ShowPracticeQuestions.tsx [hunk 6]
```diff
@@ -105,21 +107,29 @@ const ShowPracticeQuestions = () => {
...
-            </QuestionsHeader>
-            <ProgressBar total={questionsCount} activeIndex={currentQuestionIndex} />
-            {
-              selectedSectionHasPracticeQuestions && currentQuestionIndex === null
-                ? <IntroScreen />
-                : <Question />
-            }
-          </QuestionsWrapper>
-        )
+        : questionsInProggress
```
Comment (PiotrKozlowski) [inline src/app/content/practiceQuestions/components/ShowPracticeQuestions.tsx:110]: You're right, this redundant code should be fixed somehow

**Review by PiotrKozlowski — CHANGES_REQUESTED** [review]

File: src/app/content/practiceQuestions/components/Question.tsx (hunk 4 above)
Comment (jarosik10) [reply src/app/content/practiceQuestions/components/Question.tsx:74]: So you mean something like this, right?
 `e.preventDefault()`
`if (...) {dispatch(finishQuestions()); return;}`
`disptach(setAnswer(...));`
or
 `e.preventDefault()`
`if (...) {dispatch(finishQuestions());}`
`else {disptach(setAnswer(...));}`

Comment (PiotrKozlowski) [reply src/app/content/practiceQuestions/components/Question.tsx:74]: yes, both options are okay
I like using `return` for this type of situations. I'm not sure if this is in this project but we have somewhere a lint rule saying no to use `else` if not necessary

File: src/app/content/practiceQuestions/components/ShowPracticeQuestions.tsx [hunk 7]
```diff
@@ -107,29 +119,11 @@ const ShowPracticeQuestions = () => {
...
-                </FormattedMessage>
-              </QuestionsHeader>
-              <ProgressBar total={questionsCount} activeIndex={currentQuestionIndex} />
-              <IntroScreen />
-            </QuestionsWrapper>
+        : questionsInProggress ? 
+        practiceQuestionsContent
+        : hasAnswers
+            ? <FinalScreen nextSection={nextSection} />
+            : practiceQuestionsContent
```
Comment (PiotrKozlowski) [inline src/app/content/practiceQuestions/components/ShowPracticeQuestions.tsx:126]: Is there a way to use `practiceQuestionsContent` only once?

**Review by PiotrKozlowski — CHANGES_REQUESTED** [review]

File: src/app/content/practiceQuestions/components/ShowPracticeQuestions.tsx (hunk 7 above)
Comment (jarosik10) [reply src/app/content/practiceQuestions/components/ShowPracticeQuestions.tsx:126]: Maybe this way:
`: hasAnswers && !questionsInProggress`
`            ? <FinalScreen nextSection={nextSection} />`
`            : practiceQuestionsContent`

File: src/app/content/practiceQuestions/components/ShowPracticeQuestions.tsx [hunk 8]
```diff
@@ -96,6 +96,20 @@ const ShowPracticeQuestions = () => {
...
+    <QuestionsWrapper>
+      <QuestionsHeader>
+        <FormattedMessage id='i18n:practice-questions:popup:questions'>
+          {(msg: string) => msg}
+        </FormattedMessage>
+      </QuestionsHeader>
+      <ProgressBar total={questionsCount} activeIndex={currentQuestionIndex} />
+      {questionsInProggress ? <Question /> : <IntroScreen />}
+    </QuestionsWrapper>
+  );
```
Comment (PiotrKozlowski) [inline src/app/content/practiceQuestions/components/ShowPracticeQuestions.tsx:112]: now you can move this component into the render

**Review by PiotrKozlowski — CHANGES_REQUESTED** [review]
Cool, it looks nice
Please fix lint errors, tests and add new ones if needed

File: src/app/content/practiceQuestions/components/ShowPracticeQuestions.tsx (hunk 8 above)
Comment (jarosik10) [reply src/app/content/practiceQuestions/components/ShowPracticeQuestions.tsx:112]: What exactly do you mean?

Comment (PiotrKozlowski) [reply src/app/content/practiceQuestions/components/ShowPracticeQuestions.tsx:112]: This shouldn't be a const anymore
Since now it is used only in one place you can replace `practiceQuestionsContent` with the jsx

File: src/app/content/practiceQuestions/components/FinalScreen.tsx [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+import { FormattedMessage } from 'react-intl';
+import { useDispatch } from 'react-redux';
+import styled from 'styled-components/macro';
+import { textRegularStyle } from '../../../components/Typography';
+import { LinkedArchiveTreeSection } from '../../types';
+import { setSelectedSection } from '../actions';
+import NextSectionMessage from './NextSectionMessage';
+
+// tslint:disable-next-line: variable-name
+const StyledFinalScreen = styled.div`
```
Comment (PiotrKozlowski) [inline src/app/content/practiceQuestions/components/FinalScreen.tsx:11]: For screens that have lower height it displays incorrectly:
image

I've had the same problem with the Empty screen and fixed it with this: https://github.com/openstax/rex-web/pull/959/files#diff-026de43244aacd548960351da3f89d97da04840b19ddef5c41ff15cf79056ac3

File: src/app/content/practiceQuestions/components/ShowPracticeQuestions.tsx [hunk 10]
```diff
@@ -83,11 +84,10 @@ export const StyledContentLink = styled.a`
 
 // tslint:disable-next-line: variable-name
 const ShowPracticeQuestions = () => {
-  const {book, page} = useSelector(contentSelectors.bookAndPage);
+  const { book, page } = useSelector(contentSelectors.bookAndPage);
```
Comment (PiotrKozlowski) [inline src/app/content/practiceQuestions/components/ShowPracticeQuestions.tsx:87]: We're trying to avoid unnecessary changes - I usually write this kind of statements with spaces, but there are many places in the code which do not use that styling

File: src/app/content/practiceQuestions/reducer.ts [hunk 11]
```diff
@@ -24,11 +24,11 @@ const reducer: Reducer<State, AnyAction> = (state = initialState, action): State
       const shouldBeOpen = action.payload.query[modalQueryParameterName] === modalUrlName
         && action.payload.action === 'PUSH';
 
-      return {...state, open: shouldBeOpen, selectedSection: null, questions: [], questionAnswers: {}};
+      return { ...state, open: shouldBeOpen, selectedSection: null, questions: [], questionAnswers: {} };
```
Comment (PiotrKozlowski) [inline src/app/content/practiceQuestions/reducer.ts:27]: You probably have some kind of extensions installed that is fixing these brackets automatically
Please disable it or configure it to use our lint config - this will be giving a lot of diff noise in large files and is making review harder

File: src/app/content/practiceQuestions/reducer.ts [hunk 12]
```diff
@@ -39,13 +39,22 @@ const reducer: Reducer<State, AnyAction> = (state = initialState, action): State
...
+        questionAnswers: {},
+        questions: [],
+        selectedSection: action.payload,
+      };
     case getType(actions.nextQuestion):
-      return {...state, currentQuestionIndex: state.currentQuestionIndex === null ? 0 : state.currentQuestionIndex + 1};
+      return {
+        ...state,
+        currentQuestionIndex: state.currentQuestionIndex === null ? 0 : state.currentQuestionIndex + 1,
+      };
```
Comment (PiotrKozlowski) [inline src/app/content/practiceQuestions/reducer.ts:55]: this change is not necessary

File: src/app/content/practiceQuestions/selectors.ts [hunk 13]
```diff
@@ -76,3 +76,13 @@ export const isFinalQuestion = createSelector(
   currentQuestionIndex,
   (count, index) => count - 1 === index
 );
+
+export const questionsInProggress = createSelector(
+  currentQuestionIndex,
+  (index) => index !== null
+);
```
Comment (PiotrKozlowski) [inline src/app/content/practiceQuestions/selectors.ts:83]: I'm not really sure if this is needed. We can just check if `currentQuestionIndex !== null` in the component if we has to

File: src/app/content/practiceQuestions/components/ShowPracticeQuestions.tsx [hunk 14]
```diff
@@ -96,6 +96,8 @@ const ShowPracticeQuestions = () => {
     const currentSectionId = section ? section.id : page ? page.id : null;
     return currentSectionId ? getNextPageWithPracticeQuestions(currentSectionId, locationFilters, book) : undefined;
   }, [book, page, section, locationFilters]);
+  const questionsInProggress = useSelector(pqSelectors.questionsInProggress);
```
Comment (PiotrKozlowski) [inline src/app/content/practiceQuestions/components/ShowPracticeQuestions.tsx:99]: We probably don' need this one

**Review by PiotrKozlowski — CHANGES_REQUESTED** [review]

File: src/app/content/practiceQuestions/reducer.ts (hunk 12 above)
Comment (jarosik10) [reply src/app/content/practiceQuestions/reducer.ts:55]: Its about the tslint max-line-length. Should I disable tslint error for such lines?

File: src/app/content/practiceQuestions/selectors.ts (hunk 13 above)
Comment (jarosik10) [reply src/app/content/practiceQuestions/selectors.ts:77]: That's true. I've made a separate selector out of it, because it was one of the goals in the description of the [issue](https://github.com/openstax/unified/issues/1340).

File: src/app/content/practiceQuestions/reducer.ts (hunk 12 above)
Comment (PiotrKozlowski) [reply src/app/content/practiceQuestions/reducer.ts:55]: error is there because you've added black spaces before/after curly brackets
if you remove them then there will be no error

File: src/app/content/practiceQuestions/selectors.ts (hunk 13 above)
Comment (PiotrKozlowski) [reply src/app/content/practiceQuestions/selectors.ts:77]: Let's leave it then

File: src/app/content/practiceQuestions/components/ShowPracticeQuestions.tsx (hunk 14 above)
Comment (PiotrKozlowski) [reply src/app/content/practiceQuestions/components/ShowPracticeQuestions.tsx:99]: we don't but it is looking cleaner with it

File: src/app/content/practiceQuestions/reducer.ts (hunk 11 above)
Comment (jarosik10) [reply src/app/content/practiceQuestions/reducer.ts:27]: Format document command in VSC does it. I will try to configure it, that it will meet the project requirements.

File: src/app/content/practiceQuestions/components/ShowPracticeQuestions.spec.tsx [hunk 15]
```diff
@@ -151,4 +154,25 @@ describe('ShowPracticeQuestions', () => {
...
+    store.dispatch(setSelectedSection(linkedArchiveTreeSection));
+    store.dispatch(receivePracticeQuestionsSummary({
+      countsPerSource: { [linkedArchiveTreeSection.id]: 3 },
+    }));
+
+    store.dispatch(setQuestions([{ id: 'asd' } as any as PracticeQuestion]));
+    store.dispatch(setAnswer({ questionId: 'asd', answer: { id: 'qwe' } as any as PracticeAnswer }));
+
+    store.dispatch(nextQuestion());
+    const component = renderer.create(render());
```
Comment (PiotrKozlowski) [inline src/app/content/practiceQuestions/components/ShowPracticeQuestions.spec.tsx:169]: please group `store.dispatch` calls together 
I also think that you could `store.dispatch(finishQuestions())` instead of dispatching `nextQuetions`

**Review by PiotrKozlowski — CHANGES_REQUESTED** [review]
Looks really good.
I've left two comments and I will approve after you deal with this formatting issue

File: src/app/content/practiceQuestions/components/FinalScreen.tsx (hunk 9 above)
Comment (PiotrKozlowski) [reply src/app/content/practiceQuestions/components/FinalScreen.tsx:11]: This is still not fully fixed
image
image

**Review by PiotrKozlowski — CHANGES_REQUESTED** [review]

**Review by PiotrKozlowski — APPROVED** [review]
